### PR TITLE
Make cluster sharding DC aware, #23231

### PR DESCRIPTION
**Pull request by patriknw**
* Sharding only within own team (coordinator is singleton)
* the ddata Replicator used by Sharding must also be only within own team
* added support for Set of roles in ddata Replicator so that can be used
  by sharding to specify role + team
* Sharding proxy can route to sharding in another team

Refs #23231

### CODE REVIEW

**Comment by patriknw** [comment]
todo: doc notes

File: akka-distributed-data/src/main/scala/akka/cluster/ddata/Replicator.scala [hunk 1]
```diff
@@ -1099,9 +1134,9 @@ final class Replicator(settings: ReplicatorSettings) extends Actor with ActorLog
   override def preStart(): Unit = {
     if (hasDurableKeys)
       durableStore ! LoadAll
-    val leaderChangedClass = if (role.isDefined) classOf[RoleLeaderChanged] else classOf[LeaderChanged]
+    // not using LeaderChanged/RoleLeaderChanged because here we need one node indepentent of team
```
Comment (gosubpl) [inline akka-distributed-data/src/main/scala/akka/cluster/ddata/Replicator.scala]: independent ;)

File: akka-distributed-data/src/main/scala/akka/cluster/ddata/Replicator.scala [hunk 2]
```diff
@@ -98,8 +101,8 @@ object ReplicatorSettings {
 }
 
 /**
- * @param role Replicas are running on members tagged with this role.
- *   All members are used if undefined.
+ * @param roles Replicas are running on members tagged with these roles.
+ *   The memember must have all given roles. All members are used if empty.
```
Comment (gosubpl) [inline akka-distributed-data/src/main/scala/akka/cluster/ddata/Replicator.scala]: member :)

File: akka-distributed-data/src/main/scala/akka/cluster/ddata/Replicator.scala [hunk 3]
```diff
@@ -1113,7 +1148,7 @@ final class Replicator(settings: ReplicatorSettings) extends Actor with ActorLog
     clockTask.cancel()
   }
 
-  def matchingRole(m: Member): Boolean = role.forall(m.hasRole)
+  def matchingRole(m: Member): Boolean = roles.forall(m.hasRole)
```
Comment (patriknw) [inline akka-distributed-data/src/main/scala/akka/cluster/ddata/Replicator.scala]: use `subsetOf`

**Comment by patriknw** [comment]
rebased, can be merged after validation

**Review by johanandren — APPROVED** [review]

**Pull request merged**